### PR TITLE
Translate parameterized gates only in gradient calculations

### LINKED ISSUE #9054: Support unparameterized opaque gates in the gradients
### What should we add?

If a gate is opaque but unparameterized, the estimator and sampler gradients can still construct the gradient circuits. Currently, this is not supported as e.g. the LCU gradients internally attempt to unroll the whole circuit to a fixed basis set. This unrolling can be restricted to only parameterized gates and (1) enable handling custom, opaque gates and (2) doesn't prevent any optimizations the compiler might do on abstract blocks.

### PR DESCRIPTION
**Pull request by Cryoris**
### Summary

Closes #9054.

### Details and comments

All gradient now preserve unparameterized operations instead of attempting to unroll them. 
This 
* allows to evaluate gradients on custom, opaque gates that individual primitives can handle
* keeps a higher level of abstraction for optimized synthesis and compilation after the gradient circuits have been constructed.

### CODE REVIEW

**Comment by Cryoris** [comment]
Tagging @a-matsuo for review 🙂

File: qiskit/algorithms/gradients/translate_parameterized.py
Comment (t-imamichi) [inline qiskit/algorithms/gradients/translate_parameterized.py:24]: This might be useful to located in `qiskit/transpiler/passes/basis`. Is this specific to gradients?

**Comment by a-matsuo** [comment]
It looks good for me except which directory we should put the code as @t-imamichi mentioned.

File: qiskit/algorithms/gradients/translate_parameterized.py
Comment (Cryoris) [reply qiskit/algorithms/gradients/translate_parameterized.py:24]: Since it was only used in the gradients I initially put it there, but you're right it might be of more general use! I can move it to the transpiler and add some tests 🙂

Comment (Cryoris) [reply qiskit/algorithms/gradients/translate_parameterized.py]: Done in [f21dfde](https://github.com/Qiskit/qiskit-terra/pull/9067/commits/f21dfdeec3110462324afd94945f724b76814ea4) 🙂

**Comment by Cryoris** [comment]
FYI in [d567074](https://github.com/Qiskit/qiskit-terra/pull/9067/commits/d56707487bf56e17e8f1f5dba19218e8d33b4235) I also removed trailing classical registers that weren't used in the LCU gradients (+ QFI) if the `add_measurement` is `False`. With the current estimators that's not really a problem but if someone develops their own primitives this could fail if they don't support classical registers.

**Comment by a-matsuo** [comment]
Thank you @Cryoris LGTM!

**Review by woodsp-ibm — APPROVED** [review]

**Review by t-imamichi — APPROVED** [review]

File: qiskit/transpiler/passes/basis/translate_parameterized.py
Comment (mtreinish) [inline qiskit/transpiler/passes/basis/translate_parameterized.py:81]: I would add support for a `Target` here too. As we're moving towards using `BackendV2`/`Target` everywhere I think it makes sense to use it where we can. This also enables you to run this per qubit with different basis gates.

File: qiskit/transpiler/passes/basis/translate_parameterized.py
Comment (mtreinish) [inline qiskit/transpiler/passes/basis/translate_parameterized.py:35]: How is this different than the unroll* passes which effectively do this already? Looking at the code it's because this only unrolls parameterized gates, but if we adding another pass to do this we should probably call out why you'd want to use this instead of any of those.

The other related question is whether we want to add this functionality to an existing pass.

File: qiskit/transpiler/passes/basis/translate_parameterized.py
Comment (mtreinish) [inline qiskit/transpiler/passes/basis/translate_parameterized.py:141]: Instead of using a converter like this why not just create the dagcircuit directly and add the op to it?

**Review by mtreinish — COMMENTED** [review]
I only looked at the transpiler pass, but I left a few comments inline. The biggest thing for me is we should add `Target` support from the start to this pass, especially because the `BasisTranslator` which is used internally already supports working with the `Target`.

File: qiskit/transpiler/passes/basis/translate_parameterized.py
Comment (Cryoris) [reply qiskit/transpiler/passes/basis/translate_parameterized.py:141]: Yep, we can do that -- I'm not so well-versed in using the DAGCircuit 😬

File: qiskit/transpiler/passes/basis/translate_parameterized.py
Comment (Cryoris) [reply qiskit/transpiler/passes/basis/translate_parameterized.py]: Added in [dc1bf75](https://github.com/Qiskit/qiskit-terra/commit/dc1bf75e1491013d790e1c7f1b4db9bd6fc78dba) (hopefully correctly lol) 😄

File: qiskit/transpiler/passes/basis/translate_parameterized.py
Comment (mtreinish) [inline qiskit/transpiler/passes/basis/translate_parameterized.py:116]: In other passes we generally just have the target take precedence over the list of name args. This is because in the typical preset pass manager construction we have to extract the list of names from the target for compatibility with passes not target aware and we often pass both `supported_gates` and `target` at the same time.

File: qiskit/transpiler/passes/basis/translate_parameterized.py
Comment (mtreinish) [inline qiskit/transpiler/passes/basis/translate_parameterized.py:135]: We should probably have a check here that calls [`target_supported()`](https://qiskit.org/documentation/stubs/qiskit.transpiler.Target.instruction_supported.html#qiskit.transpiler.Target.instruction_supported) if the target is set. Part of what makes the target different from a list of globally supported instructions is that the gates are defined per qubit (and they're also parameter aware, so you can have multiple fixed angle variants of a gate defined for a subset of a qubits).

**Review by mtreinish — COMMENTED** [review]
Thanks for adding the target support, just two small inline comments about it

File: qiskit/transpiler/passes/basis/translate_parameterized.py
Comment (Cryoris) [reply qiskit/transpiler/passes/basis/translate_parameterized.py]: Done in [3a998f4](https://github.com/Qiskit/qiskit-terra/pull/9067/commits/3a998f4be383e7b81f5bba5ff8ce7d35d8cf5195)

File: qiskit/transpiler/passes/basis/translate_parameterized.py
Comment (Cryoris) [reply qiskit/transpiler/passes/basis/translate_parameterized.py]: Done in [3a998f4](https://github.com/Qiskit/qiskit-terra/pull/9067/commits/3a998f4be383e7b81f5bba5ff8ce7d35d8cf5195). It was quite painful to map the qubits, that `node.qargs` returned, to global indices, which `Target.instruction_supported` requires. Is there a way we could simplify that and e.g. have the `DAGOpNode` return indices instead of qubits?

File: qiskit/transpiler/passes/basis/translate_parameterized.py
Comment (mtreinish) [inline qiskit/transpiler/passes/basis/translate_parameterized.py:181]: So I may have given bad advice on this, thinking more critically about this this could be potentially incorrect in the case you're running this pass before a layout is applied. If no layout is applied this pass will treat the qubit position in a virtual circuit as the physical qubit on the target is running on (i.e. q0 in the circuit is q0 on the device) and incorrectly return here. If we can assume the circuit has a layout applied whenever this is run then this is fine (basis translator makes this assumption). But if we can't make this assumption I think dropping the `qubit_indices` here is ok since we're just checking if the target supports the operation broadly and not specifically on any qubit.

In practice all current backends I'm aware of have globally supported gates so this doesn't matter too much, but it's not guaranteed that all gates are available on all qubits in a target.

**Review by mtreinish — DISMISSED** [review]
I just left one question inline, but I had a question inline and I'm not sure how you want to proceed with regards to that. I'm fine either keeping the qubit restriction on the target lookup or dropping it depend on the expectations around the pass. So I'm karking this as approved, but if you want to change that line go ahead and feel free to merge this without waiting on a re-approval from me.

**Review by mtreinish — APPROVED** [review]

**Pull request merged**
